### ISSUE TITLE
Question: Should we implement OpenID at all?

### ISSUE THREAD

**Issue by thenomadlad** [adityadhara/RA-Scheduler#5]
Mainly because it's changing pretty drastically on April 2015.

(Also because it's annoying - what if someone tries to login with their email address using OUR form? then we have to detect what OpenID provider they're attempting to use and redirect.)

I'm thinking atleast until flask gets support for the newer OID, we should just focus on our own - Sha256 powered login/pwd tables

**Comment by gbrova** [comment]
I thought there was a flask plugin we could use for openid, and the work
we'd have to do would be minimal.  This takes care of figuring out the
provider, and presumably would get updated after the changes in April. Take
a look at the workflow in the tutorial,
http://blog.miguelgrinberg.com/post/the-flask-mega-tutorial-part-v-user-logins

On the other hand, if we're confident that our own thing is secure, then
it's fine not to implement both.

**Comment by thenomadlad** [comment]
yeah - and that's what's implemented right now, but that plugin miguel demonstrated is for openid 2.0.
take a look at this: https://github.com/mitsuhiko/flask-openid/issues/36

So do you think when april comes along - they'll replace flask-oid with connect support?
In fact, whether flask-oid replaces their core or some new package flask-oid-connect comes out, my hunch is we will have to refactor code pretty heavily in april because oid-connect is based on oauth, and flask-oauth is pretty different in implementation to flask-oid.

But if you think changes in oid is a problem that's far away enough that we'll figure it out later, I'll continue working with oid in our system. otherwise, I was thinking we backup the login management with oid and have just our management (this will take me 10 mins).

**Comment by gbrova** [comment]
Okay, you're right.  Screw it, let's forget about openid and use our own
thing.  (Feel free to remove it from the code)

Incidentally I'm having some trouble with the log in right now, but I'm
supposed to be listening to a talk so I'll try to figure it out a bit
later. I'll let you know in ~1hr if I still have trouble.

**Comment by gbrova** [comment]
Discussed offline, agreed to not use openid.

**Comment by thenomadlad** [comment]
So here's the guide I used:
https://crackstation.net/hashing-security.htm

From this I added salting and another thing of my own. I probably should
change the salt in our backend to be as long as the website says, but I'm
just too lazy to implement it for now (:P). Ooh I'll add it to the list of
preproduction tasks.

_To explain in the context of our database:_
_Salt:_ salting our password hashes ensure that having access to just our
password table or just our user table (which contains the salt), won't make
it easy for the hacker to crack a password. however, If a hacker has access
to both tables, then lookup and other methods can be used to crack the
passwords. if a hacker has access to both tables, we have a lot more to
worry about too lol

_Changing salt:_ everytime a user logs in, the salt associated to his/her
id is generated again. This ensures that a single dump of our database
wouldn't be enough for a hacker to gain access to our information.
unfortunately, the hacker would still be able to crack the passwords given
enough cpu-time, but access to stuff like .

*Note: *we are not trying to protect one user from losing their password to
a hacker. If a virus/keylogger/hacker is tracking a user's computer, then
i'm assuming it's their stupidity for not using firewalls etc. when the
user is soo important. We can help them make a secure password and put a
difficult hash on it (like the JavaScript sha256 we are using on the client
side), but ultimately, we are only protecting our table of passwords so
that a hacker can't get ALL our info in one database dump.

_Note also:_ this doesn't exempt the scenario of brute force attack from a
computer (eg. I try every combination of ascii characters for email address
"aditya1992@gmail.com". we could implement something else like a maximum
5-times failure of password matching - but it's not important now though is
it?)
